### ISSUE TITLE
igv.js as a shiny widget (R+javascript): new track only visible after manual window resize

### ISSUE THREAD

**Issue by paul-shannon** [igvteam/igv.js#557]
I have wrapped igv.js to be an R/Shiny widget.  Everything works well -except- that when I load a bed (annotation) track programmatically (R sends a json message to the browser, loadTrack is called), the track does not appear.   

When I use the mouse to manually resize the web browser window, then the new track is rendered.

Is there a function in the api which triggers the same sort of redrawing, and track rendering, which I get from a manual resize?   I have tried calling "repaint", programmatic resizing of the surrounding div, zoomIn and zoomOut, but none work.  The new track is invisible until I manually resize the browser.

I'll be most grateful for quick advice!

 - Paul

**Comment by paul-shannon** [comment]
Figured this out:  igv does not render to a hidden div.  If I ensure that igv is visible - in my case, that the tab in which it is contained is active and on top - then the track is displayed as expected.

**Comment by jrobinso** [comment]
Thanks for the update.  I hidden div has a width of zero, IIRC.   IGV creates a canvas based on the width of the track viewport.   So if it's zero it can't do it.   I don't know how to emulate the tabs in plain javascript to see if there is a workaround.

**Comment by paul-shannon** [comment]
Is a jQuery tabs demo close enough to plain javascript?

**Comment by jrobinso** [comment]
Yes, that should do.     Is there some "tab show" event?  If we could subscribe to an event and get notified we could do a global repaint.  This is how window resizing is handled.

**Comment by paul-shannon** [comment]
Sure thing.  Once a Monday noon deadline passes I will put together a tiny demo using jQuery tabs.

 - Paul

**Comment by paul-shannon** [comment]
@jrobinso 

Here is a 62-line demonstration of the problem.   It should load and run as is.  The "explanation" tab describes the one line change needed to see igv render properly.

````
<!doctype html>
<head>
  <meta charset="utf-8">
  <link rel="stylesheet" href="http://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
  <script src="https://code.jquery.com/jquery-1.12.4.js"></script>
  <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>
  <script  src='http://igv.org/web/release/1.0.9/igv-1.0.9.js'></script>
  <link rel="stylesheet" href="http://igv.org/web/release/1.0.9/igv-1.0.9.css">

<script>
  $(function() {
    $("#tabs").tabs();
    var igvDiv, options;
    igvDiv = $("#igvDiv")[0];
    options =
      {locus: "AQP4",
        minimumBases: 5,
        showRuler: true,
        reference: {id: "hg38",
              fastaURL: "https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg38/hg38.fa",
           cytobandURL: "https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg38/cytoBandIdeo.txt"
           },
        tracks: [
          {name: 'Gencode v24',
                url: "https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg38/genes/gencode.v24.annotation.sorted.gtf.gz",
           indexURL: "https://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg38/genes/gencode.v24.annotation.sorted.gtf.gz.tbi",
           format: 'gtf',
           visibilityWindow: 2000000,
           displayMode: 'EXPANDED'
           },
          ]
        }; // options
        browser = igv.createBrowser(igvDiv, options);


  } );
</script>

</head>
<body>

<div id="tabs">
  <ul>
    <li><a href="#tabs-1">Explanation</a></li>
    <li><a href="#igvTab">igv</a></li>
  </ul>

  <div id="tabs-1">
    <ul>
       <li> When this tab (really, just a div) comes first in the document, then igv renders incompletely in its div
       <li> Reverse the order of the two li items in the unordered list, the igvDiv is then visible, and it renders nicely
    </ul>
  </div>
  <div id="igvTab">
    <div id="igvDiv"></div>
  </div>
</div>


</body>
</html>
`````

**Comment by jrobinso** [comment]
Paul,  try putting this in the <script> tag. 

            $("#tabs").on("tabsactivate", function (event, ui) {
                igv.browser.resize();
            });

**Comment by jrobinso** [comment]
There is a minor issue with the end position which I'll fix tomorrow (see screenshot), but I think this can be made to work.

<img width="409" alt="screen shot 2018-03-19 at 10 46 39 pm" src="https://user-images.githubusercontent.com/933148/37637922-94172214-2bc7-11e8-894d-6add2c68b0ab.png">

**Comment by crashete** [comment]
I was facing the same problem but with modals instead of tabs. 
Solved it with a workaround creating the IGV Browser object the first time the tab is shown instead of doing it when the page loads. Something similar as @jrobinso pointed out.


```
var browser;
$("#tabs").on("tabsactivate", function (event, ui) {
      if (undefined === browser){
          browser = igv.createBrowser(igvDiv, options);
      }
 });

```

Didn't knew about the resize function, might be a better option.

**Comment by jrobinso** [comment]
I'm working on this but its complicated.    The width of zero in the hidden tab caused some issues but I've fixe those.  Now the jquery-ui css stiles seem to be squashing the icons (font-awesome).   I will look at that next, hopefully tomorrow or by end of the week at the latest.
